### PR TITLE
Master text filter range lu

### PR DESCRIPTION
**Pull request by LucasLefevre**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/spreadsheet/static/src/global_filters/plugins/global_filters_core_plugin.js [hunk 1]
```diff
@@ -210,6 +266,12 @@ export class GlobalFiltersCorePlugin extends spreadsheet.CorePlugin {
      */
     import(data) {
         for (const globalFilter of data.globalFilters || []) {
+            if (globalFilter.type === "text" && globalFilter.rangeOfAllowedValues) {
+                globalFilter.rangeOfAllowedValues = this.getters.getRangeFromSheetXC(
+                    "", // there's no default sheet, global filters are cross-sheet
```
Comment (hokolomopo) [inline addons/spreadsheet/static/src/global_filters/plugins/global_filters_core_plugin.js:271]: will end up with traceback if for whatever reason there's no sheet name in the XC :x

File: addons/spreadsheet/static/src/global_filters/plugins/global_filters_ui_plugin.js [hunk 2]
```diff
@@ -262,6 +265,46 @@ export class GlobalFiltersUIPlugin extends spreadsheet.UIPlugin {
         }
     }
 
+    /**
+     * Returns the possible values a text global filter can take
+     * if the values are restricted by a range of allowed values
+     * @param {string} filterId
+     * @returns {{value: any, formattedValue: string}[]}
```
Comment (hokolomopo) [inline addons/spreadsheet/static/src/global_filters/plugins/global_filters_ui_plugin.js:272]: might as well return the values directly as strings, the only time we use them (in the side panel) we need string values, and we cast them as string

File: addons/spreadsheet/static/src/global_filters/plugins/global_filters_ui_plugin.js [hunk 3]
```diff
@@ -262,6 +265,46 @@ export class GlobalFiltersUIPlugin extends spreadsheet.UIPlugin {
...
+            uniqueValues.add(filter.defaultValue);
+            allowedValues.push({ value: filter.defaultValue, formattedValue: filter.defaultValue });
+        }
+        // the current value might not be in the range
+        // if the range cells changed in the meantime
+        const currentValue = this.getGlobalFilterValue(filterId);
+        if (currentValue && !uniqueValues.has(currentValue)) {
+            uniqueValues.add(currentValue);
+            allowedValues.push({ value: currentValue, formattedValue: currentValue });
+        }
```
Comment (hokolomopo) [inline addons/spreadsheet/static/src/global_filters/plugins/global_filters_ui_plugin.js:304]: probably possible do everything in a single loop, without 2 extra if for default and current values, if you care about writing short code.

Someing like array = cell.filter(notempty).map(toValueAndFormattedValue); array.concat(currentValue, defaultValue); ...

File: addons/spreadsheet/static/tests/global_filters/filter_value_component_test.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import { editInput, editSelect, getFixture, mount } from "@web/../tests/helpers/utils";
+import { makeTestEnv } from "@web/../tests/helpers/mock_env";
+import { registry } from "@web/core/registry";
+import { nameService } from "@web/core/name_service";
+
+function beforeEach() {
+    registry.category("services").add("name", nameService);
+}
+
+QUnit.module("FilterValue component", { beforeEach });
```
Comment (hokolomopo) [inline addons/spreadsheet/static/tests/global_filters/filter_value_component_test.js:16]: We only test FiterTextValue, not FilterValue

File: addons/spreadsheet/static/tests/utils/zones.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+/** @odoo-module */
+
+import { helpers } from "@odoo/o-spreadsheet";
+
+const { toZone } = helpers;
+
+export function toRangeData(sheetId, xc) {
+    // FIXME should be toUnboundedZone
```
Comment (hokolomopo) [inline addons/spreadsheet/static/tests/utils/zones.js:8]: FIXME

**Review by hokolomopo — COMMENTED** [review]
Sounds good :+1:

File: addons/spreadsheet/static/src/global_filters/components/filter_text_value/filter_text_value.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import { Component } from "@odoo/owl";
+import { _t } from "@web/core/l10n/translation";
+
+export class TextFilterValue extends Component {
+    static template = "spreadsheet.TextFilterValue";
+    static props = {
+        filterId: String,
+        onValueChanged: Function,
+        model: Model,
+    };
```
[the remark this answers is not in the document]
Comment (LucasLefevre) [reply addons/spreadsheet/static/src/global_filters/components/filter_text_value/filter_text_value.js:13]: it was before we could use the `static` keyword. I believe we should gradually move to using only the static version

File: addons/spreadsheet/static/src/global_filters/components/filter_text_value/filter_text_value.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
+/** @odoo-module */
```
[the remark this answers is not in the document]
Comment (hokolomopo) [reply addons/spreadsheet/static/src/global_filters/components/filter_text_value/filter_text_value.js:1]: An empty value then. Picking "Chose a value..." in the droppdown is _really_ strange :x

File: addons/spreadsheet/static/src/global_filters/plugins/global_filters_core_plugin.js (hunk 1 above)
Comment (rrahir) [reply addons/spreadsheet/static/src/global_filters/plugins/global_filters_core_plugin.js:271]: 👍  that's a caveat in the getter

File: addons/spreadsheet/static/src/global_filters/components/filter_text_value/filter_text_value.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (rrahir) [reply addons/spreadsheet/static/src/global_filters/components/filter_text_value/filter_text_value.js:1]: We already have "Select a period..." for the the date filters. Whatever the descision, I think we should stay coherent.
we could also put the "Choose a value/ Select a value" as a disabled entry in the drodown, and users would have to use the cross sign to empty the select field?

[the remark this answers is not in the document]
Comment (LucasLefevre) [reply addons/spreadsheet/static/src/global_filters/components/filter_text_value/filter_text_value.js:1]: > we could also put the "Choose a value/ Select a value" as a disabled entry in the drodown, and users would have to use the cross sign to empty the select field?

That's an alternative I tried, but it also felt weird to me. I don't know

If the decision would me solely mine, I would just do just like "Select a period..." for simplicity, even if I agree it's not great.

File: addons/spreadsheet/static/src/global_filters/plugins/global_filters_core_plugin.js (hunk 1 above)
Comment (rrahir) [reply addons/spreadsheet/static/src/global_filters/plugins/global_filters_core_plugin.js:271]: https://www.odoo.com/web#id=3578461&cids=1&menu_id=4720&action=333&active_id=2328&model=project.task&view_type=form

**Comment by rrahir** [comment]
@robodoo r+

**Pull request closed without merging**